### PR TITLE
angband: add meta.platforms, use finalAttrs pattern

### LINKED ISSUE #282926: Build failure: angband (due to missing X11 header/library); possibly easy fix?
### Steps To Reproduce

1. build angband

### Build log

```
bash> nix profile install nixpkgs#angband
error: builder for '/nix/store/9c8dp0jgyacz45lxp4qdch99n5hcnf3x-angband-4.2.5.drv' failed with exit code 2;
       last 10 log lines:
       > Successfully compiled main-gcu.c.
       > main-x11.c:110:10: fatal error: 'X11/Xlib.h' file not found
       > #include <X11/Xlib.h>
       >          ^~~~~~~~~~~~
       > 1 error generated.
       > Failed to compile main-x11.c!
       > make[3]: *** [../mk/buildsys.mk:383: main-x11.o] Error 1
       > make[2]: *** [../mk/buildsys.mk:137: all] Error 2
       > make[1]: *** [mk/buildsys.mk:144: subdirs] Error 2
       > make: *** [mk/buildsys.mk:135: all] Error 2
       For full logs, run 'nix log /nix/store/9c8dp0jgyacz45lxp4qdch99n5hcnf3x-angband-4.2.5.drv'.
```

### Additional context

macOS Sonoma 14.3

### Notify maintainers

@kenranunderscore 

### Metadata

```console
[pmarreck@nixnerds:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"aarch64-darwin"`
 - host os: `Darwin 23.3.0, macOS 14.3`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.18.1`
 - channels(pmarreck): `""`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by ghost**
## Description of changes

fixes build for darwin targets

closes #282926

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/games/angband/default.nix [hunk 1]
```diff
@@ -27,11 +43,24 @@ stdenv.mkDerivation rec {
   configureFlags = lib.optional enableSdl2 "--enable-sdl2";
 
   installFlags = [ "bindir=$(out)/bin" ];
+}
+else stdenv.mkDerivation {
```
Comment (wegank) [inline pkgs/games/angband/default.nix:47]: I don't think you need two `mkDerivations` for darwin support to happen, especially if everything is built from source. Try using

- `buildInputs = lib.optionals stdenv.isLinux [ ... ] ++ lib.optionals stdenv.isDarwin [ ... ]`
- `buildPhase = if stdenv.isDarwin then "something" else null`
- `makefile = if stdenv.isDarwin then "Makefile.osx" else "Makefile"`

to fit the logic into one.

File: pkgs/games/angband/default.nix [hunk 2]
```diff
@@ -13,25 +13,53 @@ stdenv.mkDerivation rec {
...
+      SDL2_ttf
+    ]
+    ++ lib.optionals stdenv.isDarwin [
+      darwin.apple_sdk_11_0.frameworks.Cocoa
+      darwin.stubs.setfile
+    ];
+
+  buildPhase = if stdenv.isDarwin
+    then ''
+      pushd src
```
Comment (wegank) [inline pkgs/games/angband/default.nix:48]: You need to add `runHook preBuild` and `runHook postBuild` when overriding `buildPhase`: https://github.com/NixOS/nixpkgs/blob/master/doc/stdenv/stdenv.chapter.md#phases-sec-stdenv-phases

File: pkgs/games/angband/default.nix [hunk 3]
```diff
@@ -13,25 +13,53 @@ stdenv.mkDerivation rec {
...
+      make -f Makefile.osx\
+        ARCHS=${arch}
+      popd
+    ''
+    else null;
+
+  installPhase = if stdenv.isDarwin
+    then ''
+      mkdir -p $out/Applications
```
Comment (wegank) [inline pkgs/games/angband/default.nix:57]: Same here.

File: pkgs/games/angband/default.nix [hunk 4]
```diff
@@ -13,25 +13,53 @@ stdenv.mkDerivation rec {
...
     description = "A single-player roguelike dungeon exploration game";
     maintainers = [ maintainers.kenran ];
     license = licenses.gpl2;
+    platforms = platforms.unix;
   };
+
+  # the toolchain on darwin can't build universal binaries
+  arch = if stdenv.isAarch64
+         then "arm64"
+         else "x86_64";
```
Comment (wegank) [inline pkgs/games/angband/default.nix:27]:
```suggestion
  arch = stdenv.hostPlatform.darwinArch;
```

**Comment by adamjedrzejewski** [comment]
@wegank: thank you for reviewing

File: pkgs/games/angband/default.nix [hunk 5]
```diff
@@ -1,37 +1,67 @@
...
+      runHook preInstall
+
+      mkdir -p $out/Applications
+      cp -r Angband.app $out/Applications
+
+      runHook postInstall
+    ''
+    else null;
+
+  configureFlags = lib.optional finalAttrs.enableSdl2 "--enable-sdl2";
```
Comment (wegank) [inline pkgs/games/angband/default.nix:64]:
```suggestion
  configureFlags = lib.optional enableSdl2 "--enable-sdl2";
```

File: pkgs/games/angband/default.nix [hunk 6]
```diff
@@ -1,37 +1,67 @@
 { lib, stdenv, fetchFromGitHub, autoreconfHook, ncurses5
 , enableSdl2 ? false, SDL2, SDL2_image, SDL2_sound, SDL2_mixer, SDL2_ttf
+, darwin
 }:
+stdenv.mkDerivation (finalAttrs: {
+  inherit enableSdl2;
```
Comment (wegank) [inline pkgs/games/angband/default.nix:6]:
```suggestion
```

**Comment by wegank** [comment]
The package fails to build on x86_64-darwin: probably `darwin.apple_sdk_11_0.stdenv` could help?

File: pkgs/games/angband/default.nix [hunk 7]
```diff
@@ -1,37 +1,67 @@
...
-    SDL2_sound
-    SDL2_mixer
-    SDL2_ttf
-  ];
-
-  configureFlags = lib.optional enableSdl2 "--enable-sdl2";
-
-  installFlags = [ "bindir=$(out)/bin" ];
-
   meta = with lib; {
```
Comment (dotlambda) [inline pkgs/games/angband/default.nix:18]: Please keep this at the bottom of the file. That will make the diff more readable.

File: pkgs/games/angband/default.nix [hunk 8]
```diff
@@ -1,37 +1,67 @@
...
+      SDL2_ttf
+    ]
+    ++ lib.optionals stdenv.isDarwin [
+      darwin.apple_sdk_11_0.frameworks.Cocoa
+      darwin.stubs.setfile
+    ];
+
+  buildPhase = if stdenv.isDarwin
+    then ''
+      runHook preBuild
```
Comment (dotlambda) [inline pkgs/games/angband/default.nix:41]:
```suggestion
  buildPhase = lib.optionalString stdenv.isDarwin ''
    runHook preBuild
```

File: pkgs/games/angband/default.nix [hunk 9]
```diff
@@ -1,37 +1,67 @@
...
+      # the toolchain on darwin can't build universal binaries
+      make -f Makefile.osx\
+        ARCHS=${stdenv.hostPlatform.darwinArch}
+      popd
+
+      runHook postBuild
+    ''
+    else null;
+
+  installPhase = if stdenv.isDarwin
```
Comment (dotlambda) [inline pkgs/games/angband/default.nix:53]: same here

File: pkgs/games/angband/default.nix [hunk 10]
```diff
@@ -1,37 +1,67 @@
...
-stdenv.mkDerivation rec {
   pname = "angband";
   version = "4.2.5";
 
   src = fetchFromGitHub {
     owner = "angband";
     repo = "angband";
-    rev = version;
+    rev = finalAttrs.version;
     sha256 = "sha256-XH2FUTJJaH5TqV2UD1CKKAXE4CRAb6zfg1UQ79a15k0=";
```
Comment (dotlambda) [inline pkgs/games/angband/default.nix:15]:
```suggestion
    hash = "sha256-XH2FUTJJaH5TqV2UD1CKKAXE4CRAb6zfg1UQ79a15k0=";
```

**Review by kenranunderscore — COMMENTED** [review]
I can't test anything on `*-darwin` sadly. Just a note that on my colleague's machine (M2, `aarch64-darwin`) the command `nix run nixpkgs#angband` ran successfully and started `angband`, while in the linked issue on an only slightly different machine it does not. To better understand how darwin builds work: what explains these differences?

Once the requested changes are made, we could also finally add the `meta.platforms` attribute, so that `angband` doesn't show up as "Not available on any platform" on https://search.nixos.org.

**Comment by adamjedrzejewski** [comment]
> The package fails to build on x86_64-darwin: probably `darwin.apple_sdk_11_0.stdenv` could help?

I did some extra testing.
- `master` builds fine on `x86_64-darwin` (vm),  fails on my `aarch64-darwin` (2021 M1 MacBook Pro 14')
- this pr fails to build on `x86_64-darwin`, succeeds on my `aarch64-darwin`
 
maybe there is something sketchy going on with darwin toolchain

**Comment by wegank** [comment]
Master fails to build on both `aarch64-darwin` and `x86_64-darwin` for me, though.

File: pkgs/games/angband/default.nix (hunk 8 above)
Comment (wegank) [reply pkgs/games/angband/default.nix]: `buildPhase = null` preserves the original one, while `buildPhase = ""` empties it.

**Comment by wegank** [comment]
@ofborg build angband angband.passthru.tests

**Comment by adamjedrzejewski** [comment]
I ran `nix run nixpkgs#angband` on a clean macOS 14.3 install inside a VM and it worked just fine. Albeit it has only curses interface, no Cocoa. 

It also built successfully after `nix-store --gc` and checking out latest master.

I guess we can close this issue?

**Review by wegank — APPROVED** [review]

**Comment by adamjedrzejewski** [comment]
This shouldn't be merged. Building from master works again for some reason. (although code didn't change)

Closing to make sure we are on the same page.

**Comment by wegank** [comment]
I'm suffering from the same error in #282926.

**Comment by adamjedrzejewski** [comment]
@wegank: Ok, let's keep this open to track it. I don't know why but it went away on my machine. I'll try later to reproduce it again. Could you post entire build log so we got more eyes on it?

**Comment by adamjedrzejewski** [comment]
@wegank from the log you posted

```
angband> checking for X... libraries /usr/X11/lib, headers /usr/X11/include
```

check if those paths exist on your system, if so, that may mean configure script tries to make the toolchain hold dependencies outside of the Nix ecosystem, which is  **very** bad and needs to be fixed

**Comment by wegank** [comment]
These paths do exist. I suppose the sandbox being relaxed by default on macOS isn't pleasant...

**Comment by wegank** [comment]
In any case, thanks a lot! I think I'm OK with this PR closed if you'd like to, but I'd prefer `meta.platforms` to be set, so that Hydra builds and caches the package on darwin.

**Comment by adamjedrzejewski** [comment]
I'll update the PR to include `meta.platforms` and use `finalAttrs` pattern with old build instructions - since what we have now borks out on x86_64-darwin.

**Comment by wegank** [comment]
I guess I've already made the changes in https://github.com/NixOS/nixpkgs/pull/283592/commits/115850eb98f1e93b749a268345cce8657e1a4814?

**Comment by adamjedrzejewski** [comment]
115850eb98f1e93b749a268345cce8657e1a4814 has build instructions that failed for me on x86_64-darwin

**Review by wegank — APPROVED** [review]

**Pull request merged**